### PR TITLE
android: rework NetworkChangeCallback to track all networks

### PR DESCRIPTION
**Pull request by andrew-d**
Instead of just tracking our default network, track all of them and decide upon each change which is the "best" option.

Updates tailscale/tailscale#13173

### CODE REVIEW

**Review by agottardo — APPROVED** [review]

File: android/src/main/java/com/tailscale/ipn/NetworkChangeCallback.kt [hunk 1]
```diff
@@ -6,94 +6,159 @@ import android.net.ConnectivityManager
...
+    // default network used by Tailscale will always show up with capability
+    // NOT_VPN=false, and we must filter out NOT_VPN networks to avoid routing
+    // loops.
+    connectivityManager.registerNetworkCallback(
+        networkConnectivityRequest,
         object : ConnectivityManager.NetworkCallback() {
+          override fun onAvailable(network: Network) {
+            super.onAvailable(network)
+
+            Log.d(TAG, "onAvailable: network ${network.toString()}")
```
Comment (kari-ts) [inline android/src/main/java/com/tailscale/ipn/NetworkChangeCallback.kt:51]: I learned that you don't need to explicitly call .toString() in string interpolation, you can just reference the object directly and Kotlin automatically calls .toString() for you

File: android/src/main/java/com/tailscale/ipn/NetworkChangeCallback.kt [hunk 2]
```diff
@@ -6,94 +6,159 @@ import android.net.ConnectivityManager
...
+              activeNetworks.remove(network)
+              maybeUpdateDNSConfigLocked("onLost", dns)
             }
           }
         })
   }
 
-  fun maybeUpdateDNSConfig(linkProperties: LinkProperties, dns: DnsConfig) {
+  // pickNonMetered returns the first non-metered network in the list of
+  // networks, or the first network if none are non-metered.
```
Comment (kari-ts) [inline android/src/main/java/com/tailscale/ipn/NetworkChangeCallback.kt:87]: Can there be a situation where there are multiple non-metered networks, so picking the first one might result in the wrong one being selected?

Comment (kari-ts) [reply android/src/main/java/com/tailscale/ipn/NetworkChangeCallback.kt:87]: Sorry, we discussed this earlier. Resolving.

Comment (kari-ts) [reply android/src/main/java/com/tailscale/ipn/NetworkChangeCallback.kt:87]: It is possible, but should be ok to just pick one.

**Review by kari-ts — APPROVED** [review]

File: android/src/main/java/com/tailscale/ipn/NetworkChangeCallback.kt (hunk 1 above)
Comment (andrew-d) [reply android/src/main/java/com/tailscale/ipn/NetworkChangeCallback.kt]: Huh, TIL; thanks!

**Pull request merged**